Add device based MQTT auto discovery (#31122)
home-assistant/home-assistant.github.io: source/_integrations/mqtt.markdown

- `encoding`

The component specific options are placed as mappings under the `components` key (abbreviated as `cmp`) like:

```json
{
  "dev": {
    "ids": "ea334450945afc",
    "name": "Kitchen",
    "mf": "Bla electronics",
    "mdl": "xya",
    "sw": "1.0",
    "sn": "ea334450945afc",
    "hw": "1.0rev2",
  },
  "o": {
    "name":"bla2mqtt",
    "sw": "2.1",
    "url": "https://bla2mqtt.example.com/support",
  },
  "cmp": {
    "some_unique_component_id1": {
      "platform": "sensor",
      "device_class":"temperature",
      "unit_of_measurement":"°C",
      "value_template":"{% raw %}{{ value_json.temperature}}{% endraw %}",
      "unique_id":"temp01ae_t",
    },
    "some_unique_id2": {
      "platform": "sensor",
      "device_class":"humidity",
      "unit_of_measurement":"%",
      "value_template":"{% raw %}{{ value_json.humidity}}{% endraw %}",
      "unique_id":"temp01ae_h",
    }
  },
  "state_topic":"sensorBedroom/state",
  "qos": 2,
}
```

The components id's under the `components` (`cmp`) key should are used as part of the discovery identification. A `platform` config option is required for each component config that is added to identify the component platform.

> To remove the components publish an empty string to the discovery topic. This will remove the component and clear the published discovery payload. It will also remove the device entry if there are no further references to it.

Reviewer: ```suggestion
To remove the components, publish an empty string to the discovery topic. This will remove the component and clear the published discovery payload. It will also remove the device entry if no further references exist.
```

> To remove a single component from the device discovery an empty config can be published as an update. Note that it is still required to add the platform option.

Reviewer: ```suggestion
An empty config can be published as an update to remove a single component from the device discovery. Note that adding the platform option is still required.
```

      "0AFFD2"
    ]
  },
  "o": {
    "name": "foobar"
  },
  "payload": "short_press",
  "topic": "foobar/triggers/button1",
  "type": "button_short_press",
  "subtype": "button_1"
}
```

**Device discovery with migration support:**
Discovery topic device: `homeassistant/device/0AFFD2/config`
Discovery id: `0AFFD2 bla` *(`0AFFD2`from discovery topic, `bla`: The key under `cmp` in the discovery payload)*
Discovery payload device:

```json
{
  "device": {
    "identifiers": [
      "0AFFD2"
    ]
  },
  "o": {
    "name": "foobar"
  },
  "cmp": {
    "bla": {
      "automation_type": "trigger",
      "payload": "short_press",
      "topic": "foobar/triggers/button1",
      "type": "button_short_press",
      "subtype": "button_1",
      "platform": "device_automation"
    }
  }
}
```

If the new device discovery payload has the same `discovery_id` and comes after the single discovery payload. Home Assistant will publish `None` (retained) to the single discovery payload to remove it. This avoids rediscovery when Home Assistant restarts.

#### Single component discovery payload

The `<component>` part in the discovery topic must be one of the supported MQTT platforms.

> The options in the payload are only used to set up one specific component. If there are more components, more discovery payloads need to be send for the other components, and it is then recommended to use [device based discovery](/integrations/mqtt/#device-discovery-payload) instead.

Reviewer: ```suggestion
The options in the payload are only used to set up one specific component. If there are more components, more discovery payloads need to be sent for the other components, and it is then recommended to use [device based discovery](/integrations/mqtt/#device-discovery-payload) instead.
```

{% enddetails %}

#### Adding information about the origin af a discovery message

> It is encouraged to add additional information about the origin that supplies MQTT entities via MQTT discovery by adding the `origin` option (can be abbreviated to `o`) to the discovery payload. For device based discovery adding shared `origin` info is required. Note that these options also support abbreviations. Information of the origin will be logged to the core event log when an item is discovered or updated.

Reviewer: ```suggestion
It is encouraged to add additional information about the origin that supplies MQTT entities via MQTT discovery by adding the `origin` option (can be abbreviated to `o`) to the discovery payload. For device-based discovery, adding shared `origin` info is required. Note that these options also support abbreviations. The origin information will be in the core event log when an item is discovered or updated.
```
